sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/NativeMethodsShouldBeWrapped.cs

```diff
@@ -0,0 +1,131 @@
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+using SonarAnalyzer.Common;
+using SonarAnalyzer.Helpers;
+
+namespace SonarAnalyzer.Rules.CSharp
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    [Rule(DiagnosticId)]
+    public sealed class NativeMethodsShouldBeWrapped : SonarDiagnosticAnalyzer
+    {
+        internal const string DiagnosticId = "S4200";
+        private const string MessageFormat = "{0}";
+        private const string MakeThisMethodPrivateMessage = "Make this native method private and provide a wrapper.";
+        private const string MakeThisWrapperLessTrivialMessage = "Make this wrapper for native method '{0}' less trivial.";
+
+        private static readonly DiagnosticDescriptor rule =
+            DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId, MessageFormat, RspecStrings.ResourceManager);
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(rule);
+
+        protected override void Initialize(SonarAnalysisContext context)
+        {
+            context.RegisterSyntaxNodeAction(ReportPublicExternalMethods, SyntaxKind.MethodDeclaration);
+            context.RegisterSyntaxNodeAction(ReportTrivialWrappers, SyntaxKind.MethodDeclaration);
+        }
+
+        private static void ReportPublicExternalMethods(SyntaxNodeAnalysisContext c)
+        {
+            var methodDeclaration = (MethodDeclarationSyntax)c.Node;
+
+            if (methodDeclaration.Modifiers.Any(IsExternal) &&
+                methodDeclaration.Modifiers.Any(IsPublicOrProtected))
+            {
+                c.ReportDiagnosticWhenActive(Diagnostic.Create(rule, methodDeclaration.Identifier.GetLocation(),
+                    MakeThisMethodPrivateMessage));
+            }
+        }
+
+        private static void ReportTrivialWrappers(SyntaxNodeAnalysisContext c)
+        {
+            var methodDeclaration = (MethodDeclarationSyntax)c.Node;
+            var methodSymbol = c.SemanticModel.GetDeclaredSymbol(methodDeclaration);

```

Comment:
Check for null.

Reply:
What can be null here?

Reply:
`methodSymbol`. It is used in `GetExternalMethods()` without a check: `methodSymbol.ContainingType`

Reply:
Should be fine now